runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkSplittableDoFnFunction.java

```diff
@@ -0,0 +1,274 @@
+    WindowedValue<KeyedWorkItem<K, V>> currentValue = iterator.next();
+    final K key = currentValue.getValue().key();
+
+    final InMemoryStateInternals<K> stateInternals = InMemoryStateInternals.forKey(key);
+
+    // Used with Batch, we know that all the data is available for this key. We can't use the
+    // timer manager from the context because it doesn't exist. So we create one and advance
+    // time to the end after processing all elements.
+    final InMemoryTimerInternals timerInternals = new InMemoryTimerInternals();
+    timerInternals.advanceProcessingTime(Instant.now());
+    timerInternals.advanceSynchronizedProcessingTime(Instant.now());
+
+    FlinkSideInputReader sideInputReader = new FlinkSideInputReader(
+        sideInputs,
+        runtimeContext);
+
+    List<TupleTag<?>> additionalOutputTags = Lists.newArrayList(outputMap.keySet());
+
+    StateInternalsFactory<String> stateInternalsFactory =
+        k -> {
+          // this will implicitly be keyed by the key of the incoming
+          // element or by the key of a firing timer
+          return (StateInternals) stateInternals;
+        };
+    TimerInternalsFactory<String> timerInternalsFactory =
+        k -> {
+          // this will implicitly be keyed like the StateInternalsFactory
+          return timerInternals;
+        };
+
+    ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(
+        Executors.defaultThreadFactory());
+
+    ((SplittableParDoViaKeyedWorkItems.ProcessFn) dofn)
+        .setStateInternalsFactory(stateInternalsFactory);
+
+    ((SplittableParDoViaKeyedWorkItems.ProcessFn) dofn)
+        .setTimerInternalsFactory(timerInternalsFactory);
+
+    ((SplittableParDoViaKeyedWorkItems.ProcessFn) dofn).setProcessElementInvoker(
+        new OutputAndTimeBoundedSplittableProcessElementInvoker<>(
+            dofn,
+            serializedOptions.get(),
+            new OutputWindowedValue<OutputT>() {
+              @Override
+              public void outputWindowedValue(
+                  OutputT output,
+                  Instant timestamp,
+                  Collection<? extends BoundedWindow> windows,
+                  PaneInfo pane) {
+                outputManager.output(
+                    mainOutputTag,
+                    WindowedValue.of(output, timestamp, windows, pane));
+              }
+
+              @Override
+              public <AdditionalOutputT> void outputWindowedValue(
+                  TupleTag<AdditionalOutputT> tag,
+                  AdditionalOutputT output,
+                  Instant timestamp,
+                  Collection<? extends BoundedWindow> windows,
+                  PaneInfo pane) {
+                outputManager.output(tag, WindowedValue.of(output, timestamp, windows, pane));
+              }
+            },
+            sideInputReader,
+            executorService,
+            10000,
```

Comment:
Is there any reason to limit the output and duration in batch? Maybe we can just run the whole SDF.

Reply:
No reason, if there's work left it will just call again when the processing-time timer fires. I'll set this to `Integer.MAX_VALUE, Duration.millis(Long.MAX_VALUE)`, though.

Reply:
Yeah that should be better in batch because we're optimizing for throughput rather than latency so bigger bundles are better. Depending on how Flink streaming runner works, it may be reasonable to choose different parameter values for streaming runner too, but the default ones may also be good.